parsl/executors/high_throughput/mpi_resource_management.py

```diff
@@ -0,0 +1,222 @@
+import logging
+import multiprocessing
+import os
+import pickle
+import queue
+import subprocess
+from enum import Enum
+from typing import Dict, List
+
+from parsl.multiprocessing import SpawnContext
+from parsl.serialize import (pack_res_spec_apply_message,
+                             unpack_res_spec_apply_message)
+
+logger = logging.getLogger(__name__)
+
+
+class Scheduler(Enum):
+    Unknown = 0
+    Slurm = 1
+    PBS = 2
+    Cobalt = 3
+
+
+def get_slurm_hosts_list() -> List[str]:
+    """Get list of slurm hosts from scontrol"""
+    cmd = "scontrol show hostname $SLURM_NODELIST"
+    b_output = subprocess.check_output(
+        cmd, stderr=subprocess.STDOUT, shell=True
+    )  # bytes
+    output = b_output.decode().strip().split()
+    return output
+
+
+def get_pbs_hosts_list() -> List[str]:
+    """Get list of PBS hosts from envvar: PBS_NODEFILE"""
+    nodefile_name = os.environ["PBS_NODEFILE"]
+    with open(nodefile_name) as f:
+        return [line.strip() for line in f.readlines()]
+
+
+def get_cobalt_hosts_list() -> List[str]:
+    """Get list of COBALT hosts from envvar: COBALT_NODEFILE"""
+    nodefile_name = os.environ["COBALT_NODEFILE"]
+    with open(nodefile_name) as f:
+        return [line.strip() for line in f.readlines()]
+
+
+def get_nodes_in_batchjob(scheduler: Scheduler) -> List[str]:
+    """Get nodelist from all supported schedulers"""
+    nodelist = []
+    if scheduler == Scheduler.Slurm:
+        nodelist = get_slurm_hosts_list()
+    elif scheduler == Scheduler.PBS:
+        nodelist = get_pbs_hosts_list()
+    elif scheduler == Scheduler.Cobalt:
+        nodelist = get_cobalt_hosts_list()
+    return nodelist
+
+
+def identify_scheduler() -> Scheduler:
+    """Use envvars to determine batch scheduler"""
+    if os.environ.get("SLURM_NODELIST"):
+        return Scheduler.Slurm
+    elif os.environ.get("PBS_NODEFILE"):
+        return Scheduler.PBS
+    elif os.environ.get("COBALT_NODEFILE"):
+        return Scheduler.Cobalt
+    else:
+        return Scheduler.Unknown
+
+
+class MPIResourceUnavailable(Exception):
+    """Raised if there are no free resources available for an MPI request"""
```

Comment:
does "resource" mean a rank? or a node? or some other abstract resource?

Reply:
In this case it is always nodes. I'll rename `s/resource/node` since that would be clearer.